### PR TITLE
Pool collections during tagging.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Followup to https://github.com/dotnet/roslyn/pull/73703.

Drops memory allocations from:

image

to just 

image

A dropping from 98mb of allocs in the ide tagger to just 20 (an 80% decrease).

### CODE REVIEW

File: src/EditorFeatures/Core/Tagging/AbstractAsynchronousTaggerProvider.TagSource.cs [hunk 1]
```diff
@@ -171,6 +172,7 @@ private sealed partial class TagSource
             _dataSource = dataSource;
             _asyncListener = asyncListener;
             _nonFrozenComputationCancellationSeries = new(_disposalTokenSource.Token);
+            _tagSpanSetPool = new ObjectPool<HashSet<ITagSpan<TTag>>>(() => new HashSet<ITagSpan<TTag>>(this), trimOnFree: false);
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Tagging/AbstractAsynchronousTaggerProvider.TagSource.cs:175]: needs to be an instance member as it creates hashsets that use 'this' as the special IEqualityComparer to compare tag-spans.

File: src/EditorFeatures/Core/Tagging/AbstractAsynchronousTaggerProvider.TagSource_IEqualityComparer.cs [hunk 2]
```diff
@@ -11,8 +12,18 @@ internal abstract partial class AbstractAsynchronousTaggerProvider<TTag>
 {
     private partial class TagSource : IEqualityComparer<ITagSpan<TTag>>
     {
+        private readonly ObjectPool<HashSet<ITagSpan<TTag>>> _tagSpanSetPool;
+
         public bool Equals(ITagSpan<TTag>? x, ITagSpan<TTag>? y)
-            => x != null && y != null && x.Span == y.Span && _dataSource.TagEquals(x.Tag, y.Tag);
+        {
+            if (x == y)
+                return true;
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Tagging/AbstractAsynchronousTaggerProvider.TagSource_IEqualityComparer.cs:20]: this clause was added as a fast path.  The rest was just broken into separate clauses.

File: src/EditorFeatures/Core/Tagging/AbstractAsynchronousTaggerProvider.TagSource_ProduceTags.cs [hunk 3]
```diff
@@ -389,52 +389,56 @@ private static void CheckSnapshot(ITextSnapshot snapshot)
 
         private ImmutableDictionary<ITextBuffer, TagSpanIntervalTree<TTag>> ComputeNewTagTrees(ImmutableDictionary<ITextBuffer, TagSpanIntervalTree<TTag>> oldTagTrees, TaggerContext<TTag> context)
         {
-            // Ignore any tag spans reported for any buffers we weren't interested in.
-
-            var spansToTag = context.SpansToTag;
-            var buffersToTag = spansToTag.Select(dss => dss.SnapshotSpan.Snapshot.TextBuffer).ToSet();
-            var newTagsByBuffer =
-                context.TagSpans.Where(ts => buffersToTag.Contains(ts.Span.Snapshot.TextBuffer))
-                                .ToLookup(t => t.Span.Snapshot.TextBuffer);
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Tagging/AbstractAsynchronousTaggerProvider.TagSource_ProduceTags.cs:398]: lots of wasteful linq/enumerator code here.  it all got changes to simple code using pools.

**Comment by CyrusNajmabadi** [comment]
@ToddGrun this is ready for review.

File: src/EditorFeatures/Core/Shared/Tagging/Utilities/TagSpanIntervalTree.cs [hunk 4]
```diff
@@ -24,13 +24,14 @@ namespace Microsoft.CodeAnalysis.Editor.Shared.Tagging;
 internal sealed partial class TagSpanIntervalTree<TTag>(
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Shared/Tagging/Utilities/TagSpanIntervalTree.cs:24]: note: my plan fo rthis type is to get rid of it, inlining its underlying IntervalTree into the tagger directly.  The helpers we have here would then be static helper methods in the tagger as they're all tagger specific helpers (most of which can be improved).

**Comment by CyrusNajmabadi** [comment]
@ToddGrun this isready for review.

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/ObjectPools/Extensions.cs [hunk 5]
```diff
@@ -152,7 +152,7 @@ public static void ClearAndFree<T>(this ObjectPool<HashSet<T>> pool, HashSet<T>
         var count = set.Count;
         set.Clear();
 
-        if (count > Threshold)
+        if (count > Threshold && pool.TrimOnFree)
```
Comment (ToddGrun) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/ObjectPools/Extensions.cs:155]: >if (count > Threshold && pool.TrimOnFree) [](http://example.com/codeflow?start=8&length=41)

Should the other ClearAndFree methods be doing this too?

Comment (CyrusNajmabadi) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/ObjectPools/Extensions.cs:155]: probably :)  i can try to address in followup pr.

File: src/EditorFeatures/Core/Tagging/AbstractAsynchronousTaggerProvider.TagSource_ProduceTags.cs [hunk 6]
```diff
@@ -443,35 +447,40 @@ private static void CheckSnapshot(ITextSnapshot snapshot)
...
-                ? oldTagTree.GetSpans(snapshot)
-                : GetNonIntersectingTagSpans(spansToInvalidate, oldTagTree);
-
-            // Then union those with the new tags to produce the final tag tree.
-            var finalTags = oldTagsToKeep.Concat(newTags);
-            return new TagSpanIntervalTree<TTag>(textBuffer, _dataSource.SpanTrackingMode, finalTags);
+            if (noSpansToInvalidate)
+            {
+                // If we have no spans to invalidate, then we can just keep the old tags and add the new tags.
+                var oldTagsToKeep = oldTagTree.GetSpans(newTags.First().Span.Snapshot);
```
Comment (ToddGrun) [inline src/EditorFeatures/Core/Tagging/AbstractAsynchronousTaggerProvider.TagSource_ProduceTags.cs:455]: >.First() [](http://example.com/codeflow?start=63&length=8)

nit: [0]

**Review by ToddGrun — APPROVED** [review]
:shipit:

File: src/EditorFeatures/Core/Tagging/AbstractAsynchronousTaggerProvider.TagSource_ProduceTags.cs (hunk 6 above)
Comment (CyrusNajmabadi) [reply src/EditorFeatures/Core/Tagging/AbstractAsynchronousTaggerProvider.TagSource_ProduceTags.cs:455]: ArrayBuilder.First() is fine.  it's not linq :)

**Pull request merged**